proposals/ref-unsafe-in-iterators-async.md

```diff
@@ -43,11 +43,128 @@ lock (this)
+- The `set` accessor of an iterator property or indexer (i.e., its `get` accessor is implemented via an iterator block)
+  "inherits" its safe/unsafe scope from the declaration.
+- This does not affect partial declarations without implementation as they are only signatures and cannot have an iterator body.
+- In C# 12 it is an error to have an iterator method marked with the `unsafe` modifier, but that is allowed in C# 13.
+
+For example:
+
+```cs
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+class A : System.Attribute { }
+unsafe partial class C1
+{ // unsafe context
+    [/* unsafe context */ A]
+    IEnumerable<int> M1(
+        /* unsafe context */ int*[] x)
+    { // safe context (this is the iterator block implementing the iterator)
+        yield return 1;
+    }
+    IEnumerable<int> M2()
+    { // safe context (this is the iterator block implementing the iterator)
+        unsafe
+        { // unsafe context
+            { // unsafe context (this is *not* the block implementing the iterator)
+                yield return 1; // error: `yield return` in unsafe context
+            }
+        }
+    }
+    [/* unsafe context */ A]
+    unsafe IEnumerable<int> M3(
+        /* unsafe context */ int*[] x)
+    { // safe context
+        yield return 1;
+    }
+    [/* unsafe context */ A]
+    IEnumerable<int> this[
+        /* unsafe context */ int*[] x]
+    { // unsafe context
+        get
+        { // safe context
+            yield return 1;
+        }
+        set { /* unsafe context */ }
+    }
+    [/* unsafe context */ A]
+    unsafe IEnumerable<int> this[
+        /* unsafe context */ long*[] x]
+    { // unsafe context (the iterator declaration is unsafe)
+        get
+        { // safe context
+            yield return 1;
+        }
+        set { /* unsafe context */ }
+    }
+    IEnumerable<int> M4()
+    {
+        yield return 1;
+        var lam1 = async () =>
+        { // safe context
+          // spec violation: in Roslyn, this is an unsafe context in LangVersion 12 and lower
+            await Task.Yield();
+        };
+        unsafe
+        {
+            var lam2 = () =>
+            { // unsafe context, lambda cannot be an iterator
+                yield return 1; // error: yield cannot be used in lambda
+            };
+        }
+        async void local()
+        { // safe context
+          // spec violation: in Roslyn, this is an unsafe context in LangVersion 12 and lower
+            await Task.Yield();
+        }
+        local();
+    }
+    public partial IEnumerable<int> M5() // unsafe context (inherits from parent)
+    { // safe context
+        yield return 1;
+    }
+}
+partial class C1
+{
+    public partial IEnumerable<int> M5(); // safe context (inherits from parent)
+}
+class C2
+{ // safe context
+    [/* unsafe context */ A]
+    unsafe IEnumerable<int> M(
+        /* unsafe context */ int*[] x)
+    { // unsafe context (the iterator declaration is unsafe)
```

Comment:
>// unsafe context (the iterator declaration is unsafe) [](http://example.com/codeflow?start=6&length=54)

This contradicts the following statement from above: "If an iterator declaration is marked with the `unsafe` modifier, the signature is in an unsafe scope
  but the iterator block used to implement that iterator still defines a safe scope."